core/include/emulate.h

```diff
@@ -0,0 +1,212 @@
+
+typedef struct em_vcpu_ops_t {
+    uint64_t (*read_gpr)(void *vcpu, uint32_t reg_index,
+                         uint32_t size);
+    void (*write_gpr)(void *vcpu, uint32_t reg_index,
+                      uint64_t value, uint32_t size);
+    uint64_t (*read_rflags)(void *vcpu);
+    void (*write_rflags)(void *vcpu, uint64_t value);
+    uint64_t (*get_segment_base)(void *vcpu, uint32_t segment);
+    void (*advance_rip)(void *vcpu, uint64_t len);
+    em_status_t (*read_memory)(void *vcpu, uint64_t ea,
+                               uint64_t *value, uint32_t size);
+    em_status_t (*read_memory_post)(void *vcpu,
+                               uint64_t *value, uint32_t size);
+    em_status_t (*write_memory)(void *vcpu, uint64_t ea,
+                                uint64_t *value, uint32_t size);
+} em_vcpu_ops_t;
+
+typedef void (em_operand_decoder_t)(struct em_context_t *ctxt,
+                                    struct em_operand_t *op);
+
+typedef struct em_opcode_t {
+    union {
+        void *handler;
+        const struct em_opcode_t *group;
+    };
+    em_operand_decoder_t *decode_dst;
+    em_operand_decoder_t *decode_src1;
+    em_operand_decoder_t *decode_src2;
+    uint64_t flags;
+} em_opcode_t;
+
+/* Context */
+typedef struct em_operand_t {
+    uint32_t size;
+    uint32_t flags;
+    em_operand_type_t type;
+    union {
+        struct operand_mem_t {
+            uint64_t ea;
+            uint32_t seg;
+        } mem;
+        struct operand_reg_t {
+            uint32_t index;
+        } reg;
+    };
+    uint64_t value;
+} em_operand_t;
+
+typedef struct em_context_t {
+    void *vcpu;
+    const struct em_vcpu_ops_t *ops;
+    bool finished;
+
+    em_mode_t mode;
+    const uint8_t *insn;
+    uint32_t operand_size;
+    uint32_t address_size;
+    uint32_t len;
+    uint64_t rip;
+
+    int rep;
+    int lock;
+    int override_segment;
+    int override_operand_size;
+    int override_address_size;
+
+    struct em_opcode_t opcode;
+    struct em_operand_t dst;
+    struct em_operand_t src1;
+    struct em_operand_t src2;
+    uint64_t rflags;
+
+    /* Decoder */
+    union {
+        struct {
+            uint8_t b : 1;
+            uint8_t x : 1;
+            uint8_t r : 1;
+            uint8_t w : 1;
+            uint8_t   : 4;
+        };
+        uint8_t value;
+    } rex;
+    union {
+        struct {
+            uint8_t rm  : 3;
+            uint8_t reg : 3;
+            uint8_t mod : 2;
+        };
+        struct {
+            uint8_t     : 3;
+            uint8_t opc : 3;
+            uint8_t     : 2;
+        };
+        uint8_t value;
+    } modrm;
+    union {
+        struct {
+            uint8_t base  : 3;
+            uint8_t index : 3;
+            uint8_t scale : 2;
+        };
+        uint8_t value;
+    } sib;
+} em_context_t;
+
+em_status_t __stdcall em_decode_insn(struct em_context_t *ctxt, const uint8_t *insn);
```

Comment:
The `__stdcall` decoration introduces in a build warning on macOS:

```
warning: calling convention '__stdcall' ignored for this target [-Wignored-attributes]
```

So basically, do we have to decorate any C function that calls any of the ASM routines? Why?

Reply:
The problem is that *haxlib.vcxproj* and *IntelHaxm.vcxproj* are stdcall by default, and *tests.vcxproj* is cdecl by default.

As mentioned earlier, the solution would be to explicitly stating the correct calling convention. However, changing *tests.vcxproj* to stdcall causes an issue with *Google Test*, and changing haxlib/IntelHaxm to cdecl causes an issue with MASM code.

The only solution left, was explicitly telling tests that these two functions are stdcall, but that's a suboptimal fix. I'm still thinking how to properly deal with this situation.

Reply:
Thanks. IIUC, your other proposal about replacing existing MASM code with NASM should be a proper fix. But for this PR, we can accept the current solution. The only issue is the macOS build warning, which I think can be fixed by something like:

```
#ifdef __i386__
#define EMCALL __stdcall
#else  // !__i386__
#define EMCALL
#endif  // __i386__

em_status_t EMCALL em_decode_insn(..);
em_status_t EMCALL em_emulate_insn(..)
```